router/src/validation.rs

```diff
@@ -506,62 +523,59 @@ fn fetch_image(input: &str) -> Result<(Vec<u8>, String, usize, usize), Validatio
     }
 }
 
+fn image_tokens(
+    config: &Config,
+    preprocessor_config: Option<&HubPreprocessorConfig>,
+    height: usize,
+    width: usize,
+) -> String {
+    use Config::*;
+    use HubPreprocessorConfig::*;
+    match config {
+        Idefics => "<image>".to_string(),
+        Idefics2(config) => {
+            let slots = config.get_number_of_features(height, width);
+
+            const FAKE: &str = "<fake_token_around_image>";
+            const IMAGE: &str = "<image>";
+            const FAKE_LEN: usize = FAKE.len();
+            const IMAGE_LEN: usize = IMAGE.len();
+
+            let mut tokens = String::with_capacity(2 * FAKE_LEN + slots * IMAGE_LEN);
+            tokens.push_str(FAKE);
+            tokens.extend(iter::repeat(IMAGE).take(slots));
+            tokens.push_str(FAKE);
+
+            if matches!(
+                preprocessor_config,
+                Some(Idefics2Processor(Idefics2Preprocessor {
+                    do_image_splitting: true,
+                    ..
+                }))
+            ) {
+                tokens = tokens.repeat(5);
```

Comment:
Careful about the removal of contiguous `<fake>` tokens in the preprocessor.

Reply:
Done. I'm now doing it in the same way as transformers. First add the duplicate tokens, then substitute them away.

I first wanted to make the image token processing aware of preceding chunks and only add a preceding fake token if there was not a preceding image (and then only intersperse the image splits with a single fake token). But that seemed like a bit of premature optimization of the image token processing.